### ISSUE TITLE
Cache not fully refreshed after applying scope

### ISSUE THREAD

**Issue by sebastiaanluca** [JosephSilber/bouncer#367]
I did some research, but I'm still not sure if this is a Bouncer issue or a quirk in my code.

### Example

```php
$user = Auth::user();

Bouncer::scope()
            ->onlyRelations()
            ->dontScopeRoleAbilities()
            ->to('myscope');

$user->assign('myrole');

Bouncer::refreshFor($user);
```

### Expected vs actual results

```php
// These first two return false (not ok; both should return true)
$user->isA('myrole');
$user->can('abilityinmyrole');

// These return true (as expected)
Bouncer::scope()
            ->onlyRelations()
            ->dontScopeRoleAbilities()
            ->to('myscope');

$user->isA('myrole');
$user->can('abilityinmyrole');

// If I clear the entire cache for all users, these return true (as expected, but this indicates the cache was not complete cleared)
Bouncer::refresh();

$user->isA('myrole');
$user->can('abilityinmyrole');
```

### Remarks

As you can see, the results of example 1 and 3 differ but should be exactly the same, right? If it would've been an issue with the scope (in which Bouncer only checks for roles or abilities with a scope if they have one), then 1 and 3 would both return false. Yet they vary in result before and after refreshing the cache.

Does the scope somehow limit the cache refresh or shouldn't it interfere with it at all?

**Comment by JosephSilber** [comment]
It seems like your examples are missing some details, which makes it hard to follow.

1. Is the code in the **Expected vs actual results** block run right after the code in the **Example** bloc, or in a different request?

1. Does the 2nd example return the same results as the 3rd one? Then what are you illustrating with `Bouncer::refresh()`?

**Comment by sebastiaanluca** [comment]
Closing this for now as I can't reproduce it after implementing a fix for https://github.com/JosephSilber/bouncer/issues/368.

**Comment by sebastiaanluca** [comment]
Just had this issue again, so reopening in the hopes of getting some feedback to further debug this as I don't have a clear view on the cache internals of Bouncer itself.

Simplified example:

```php
// Provider setup
$bouncer = app(Bouncer::class);

$bouncer->useUserModel(User::class);
$bouncer->runAfterPolicies();
$bouncer->cache();

// Runtime code example
$user = User::first();

Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to('tenant:1');

// User model uses HasRoles trait
$user->assign('tenant_admin');

Bouncer::refreshFor($user);
```

Anything checking the **user's role in that scope** returns `false`:

```php
// Setup
Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to('tenant:1');

// In a request controller
$this->authorize('random_tenant_admin_permission');
// false

// Checking the role in a Blade view
@can('random_tenant_admin_permission')
    You can do this!
@endcan
// Nothing shows
```

If I replace `Bouncer::refreshFor($user);` with `Bouncer::refresh();`, everything works as intended and the user **is authorized** and **can see the "You can do this!" message.**

Is there anything in the example above that's set up wrongly in regards to scope in combination with refreshing the cache? If a user gets a tenant role, I don't want to clear the entire Bouncer cache, but just the one for this user (and more granularly, in this scope, but that's not a requirement).

**Comment by JosephSilber** [comment]
What does `tenant:1` mean? Have you changed the type of the `scope` column? By default it's an `integer`.

**Comment by sebastiaanluca** [comment]
`tenant:1` is just the scope and yes, I've changed it in the migration to allow string type scopes:

```
$table->string('scope')->nullable()->index;
```

So when we assign a role in a scope and refresh the cache for that "authorizer" / entity only, any checks that should succeed (in that scope) fail. If we refresh the entire cache, it works as expected.

I'd be happy to debug further if you have any instructions. Maybe it is something app-specific, but besides the above example, we're not doing anything else that's related to Bouncer.

**Comment by JosephSilber** [comment]
I created [a demo to test this issue](https://github.com/JosephSilber/bouncer-issue-367-demo), but I couldn't reproduce it. All tests seem to pass.

Look at [this commit](https://github.com/JosephSilber/bouncer-issue-367-demo/commit/fd9032b07544968721a916b54e07643094b717d0), which basically tries to recreate your setup.

---

To try it yourself:

1. Clone [the repo](https://github.com/JosephSilber/bouncer-issue-367-demo)
1. Run `composer install`
1. Create an `.env` file and add credentials to an empty database
1. Run `php artisan migrate`
1. Run `phpunit`

All tests pass.

**Comment by JosephSilber** [comment]
Since I couldn't reproduce your results, and you haven't responded, I'm closing this issue.

If you have additional information, feel free to reopen this issue with the relevant information.

**Comment by sebastiaanluca** [comment]
I'll try your demo asap and see how it corresponds with our setup. Maybe it's related to our migrations using a string, but we'll see what comes up. Thanks for the help so far!

**Comment by sebastiaanluca** [comment]
I've tested your example and added some extra tests to verify the bug in our code. I think the main culprit/difference is our requests are never run in a single scope from start to end. We don't set a scope at the start and execute everything in it, but utilize different scopes (different tenants and types of tenants) throughout a single request. Most of the time even a combination of tenant and no tenant (global).

To illustrate the difference, your example / usual use case:

```php
Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to(1);

$user->assign('tenant_admin');

$this->bouncer->refreshFor($user);

// Returns true (aka works)
$this->assertTrue($user->isA('tenant_admin'));
```

Ours:

```php
Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to(2);

$user->assign('tenant_admin');

// Remove the scope so the rest of the code is run without a tenant
Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to(null);

$this->bouncer->refreshFor($user);

// Returns false (doesn't work)
$this->assertTrue($user->isA('tenant_admin'));
```

The main take away here is that you need to refresh the cache for an object **in a scope**. Executing that code outside a scope doesn't seem to do anything, unless you clear the entire cache (which we did as a quick fix).

```php
Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to(2);

$this->bouncer->refreshFor($user);

// Returns true (works)
$this->assertTrue($user->isA('tenant_admin'));

Bouncer::scope()
    ->onlyRelations()
    ->dontScopeRoleAbilities()
    ->to(null);
```

So my final question would be: is there a way to clear the Bouncer cache for a single user/object for all scopes (when clearing it outside a scope / when no scope is active)?

**Comment by sebastiaanluca** [comment]
Quick follow-up on this issue as it seems we're still doing something wrong (?).

Everything runs fine in our tests (single request), but when assigning a role to a user and refreshing its cache in one request (POST) and checking it in another (any other following GET request), the user **has the role but not its abilities**. Only when refreshing the cache for that user once again (in addition to refreshing directly after assigning the role) or clearing the Bouncer cache entirely, does the user also get that role's abilities. All of this regardless of scopes (same result with or without). We're using Redis as cache store btw.

Outline:

Request 1 (POST):
- `Bouncer::scope()->onlyRelations()->dontScopeRoleAbilities()->to(2)`
- `$user->assign('role')`
- `Bouncer::refreshFor($user)`
- `Bouncer::scope()->onlyRelations()->dontScopeRoleAbilities()->to(null)` (clear scope)

Request 2 (GET)
- `$user->isA('role')` (always returns `true` as desired, even when scoped)
- `$user->can('role_ability')` (returns `false` while it should be `true`, even when scoped)

If we add `Bouncer::refreshFor($user);` in the second request before anything else, every check returns `true` (as desired).

Are we missing something or any way to further debug this? Could this be related to now clearing the cache inside a scope and request 2 using cache that has not been updated for use outside the scope?

**Comment by sebastiaanluca** [comment]
Can verify the abilities of a role get updated and cached correctly when refreshing **in the scope** and **outside the scope** for a single user at the same time.

For example:

```
$user->assign('admin');

Bouncer::refreshFor($user);

Bouncer::scope()->onceTo('tenant', function () use($user) {
    Bouncer::refreshFor($user);
});
```

Or simply clear the entire cache, but not so performant. @JosephSilber Can you shine your light on this? Not sure if this is the way to go as it feels a bit hacky. Would assume clearing the cache inside or outside the scope clears all permissions of a user, not just the roles or abilities.

**Comment by JosephSilber** [comment]
> Can you shine your light on this? 

I'm not sure what the question is.

**Comment by sebastiaanluca** [comment]
That refreshing the cache for a certain user should clear _all_ cached queries and data, shouldn't it? Now it only does so for data __outside scopes__. We have to manually go over each scope and clear the cache for that scope if we've updated a role, ability, permission for a user in that scope.

```php
Bouncer::refreshFor($user);
```

Should be enough to clear everything IMO.

**Comment by JosephSilber** [comment]
I think this is working as designed.

Every scope is meant to be consumed as its own context, and it doesn't make sense to cross contexts like that 🤷‍♂️

**Comment by grantholle** [comment]
I think I understand @sebastiaanluca situation as I just encountered it myself. Thanks for explaining it so well and helping me figure out what was happening.

> Every scope is meant to be consumed as its own context, and it doesn't make sense to cross contexts like that

I agree with this statement, however when it comes to _cache_ specifically, I feel like it ultimately doesn't matter that it gets refreshed across scopes. Since `cache()` is called at a non-scoped/global level, it makes sense to me that `refreshFor()` also cleans the cache entirely for a single user.

Either way, I understand what I was doing wrong and I can get it to work as expected.
